cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/common/AbstractWatcherThread.java

```diff
@@ -47,62 +66,62 @@
   private static final String DELETED = "DELETED";
   // The response type from K8s when a resource was modified
   private static final String MODIFIED = "MODIFIED";
+  // The response type from K8s when there is error from the watch
+  private static final String ERROR = "ERROR";
+
+  protected final String group;
+  protected final String version;
+  protected final String plural;
 
   private final String namespace;
   private final Random random;
   private final Type resourceType;
-  private volatile Watchable<T> watch;
+  private final AtomicReference<Watchable<DynamicKubernetesObject>> watch;
+  private final CachingResourceChangeListener<T> changeListener;
   private volatile boolean stopped;
+  private volatile ApiClient apiClient;
+
+  protected AbstractWatcherThread(String threadName, String namespace, String group, String version, String plural) {
+    this(threadName, namespace, group, version, plural, null);
+  }
 
-  protected AbstractWatcherThread(String threadName, String namespace) {
+  @VisibleForTesting
+  AbstractWatcherThread(String threadName, String namespace, String group, String version, String plural,
+                        @Nullable ApiClient apiClient) {
```

Comment:
where are we calling this constructor (passing non-null apiClient)?

Reply:
From unit-test, as marked with VisibleForTesting.